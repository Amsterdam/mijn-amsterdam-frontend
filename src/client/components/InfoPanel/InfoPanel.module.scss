@import '../../styles/_global.scss';

.InfoPanel {
  border-top: 0.1rem solid $color-neutral-grey2;
  padding-top: 3rem;
}

.Title {
  margin-left: 16rem;
  margin-bottom: 1.8rem;

  @include mq-phone() {
    margin-left: 0;
  }
}

.InfoPanelContent {
  display: flex;
  flex-wrap: wrap;
}

.InfoPanelCollapsible {

  @include select(SectionCollapsible_Title_) {
    @include heading-medium;
    @include mq-phone {
      margin-left: 0;
    }
  }

  @include select(SectionCollapsible_Panel) {
    .InfoPanelContent {
      border-top: 0;
    }

    .InfoPanel {
      margin-bottom: 2rem;
    }
  }
}

.TableWrap {
  width: 59%;
  @include mq-phone() {
    width: 100%;
  }

  + .TableWrap {
    border-top: .1rem solid $color-neutral-grey2;
    margin-top: 2rem;
    padding-top: 2rem;
  }

  th span {
    white-space: nowrap;
  }
  @include mq-desktop() {
    th {
      position: relative;
      span {
        position: absolute;
        right: 2rem;
        top: .1rem;
      }
    }
  }
}

.InfoPanelTableItem {

}

.InfoPanelTable {
  border-collapse: collapse;

  tbody > tr {
    > td {
      text-align: left;
      white-space: pre-line;
      padding-bottom: 0.5rem;
    }
    > th {
      width: 16rem;
      text-align: right;
      padding-right: 2rem;
      padding-bottom: 0.5rem;
      color: $color-neutral-grey4;
      font-weight: 200;
    }

    th,
    td {
      vertical-align: top;
    }
  }

  @include mq-phone() {
    display: block;

    tbody,
    thead,
    tr,
    td,
    th {
      display: block;
    }
    tbody > tr {
      > th,
      > td {
        width: auto;
        text-align: left;
        width: auto;
        padding-bottom: 0;
      }

      + tr {
        > th {
          padding-top: 2rem;
        }
      }
    }
  }
}

.InfoPanelActionLinks {
  @include reset-list-style;
  position: relative;

  > li {
    margin-bottom: 0.6rem;
  }

  @include mq-phone() {
    margin: 2rem 0 0 0;
  }
}
