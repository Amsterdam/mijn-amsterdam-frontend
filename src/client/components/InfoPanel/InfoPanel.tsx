import Heading from '../Heading/Heading';
import Linkd from '../Button/Button';
import React, { useMemo } from 'react';
import SectionCollapsible from '../SectionCollapsible/SectionCollapsible';
import { Unshaped } from '../../../universal/types';
import classnames from 'classnames';
import { entries } from '../../../universal/helpers';
import slug from 'slugme';
import styles from './InfoPanel.module.scss';

export interface ActionLink {
  title: string;
  url: string;
  external?: boolean;
}

export type ActionLinksFormatter =
  | ActionLink[]
  | ((sourceData: any) => ActionLink[]);

export interface InfoPanelActionLinksProps {
  actionLinks: ActionLink[];
}

export interface InfoPanelTableProps {
  panelData: Unshaped;
  omitPairWithFalseyValues?: boolean;
}

export interface InfoPanelProps {
  id?: string;
  title?: string;
  actionLinks?: ActionLink[];
  panelData: Unshaped | Unshaped[];
  className?: string;
  omitPairWithFalseyValues?: boolean;
}

export interface InfoPanelCollapsibleProps extends InfoPanelProps {
  id: string;
  startCollapsed?: boolean;
}

function InfoPanelActionLinks({ actionLinks }: InfoPanelActionLinksProps) {
  return (
    <ul className={styles.InfoPanelActionLinks}>
      {actionLinks.map(actionLink => (
        <li key={actionLink.title}>
          <Linkd href={actionLink.url} external={actionLink.external}>
            {actionLink.title}
          </Linkd>
        </li>
      ))}
    </ul>
  );
}

function getValue(value: any) {
  if (React.isValidElement(value)) {
    return value;
  }
  if (Array.isArray(value) || typeof value === 'object') {
    try {
      return JSON.stringify(value);
    } catch (error) {}
  }
  return value;
}

function filterValue(
  omitPairWithFalseyValues: boolean = true,
  [, value]: [string, string | number]
) {
  return omitPairWithFalseyValues ? !!value : true;
}

function InfoPanelTable({
  panelData = {},
  omitPairWithFalseyValues = true,
}: InfoPanelTableProps) {
  const tables = useMemo(() => {
    return Array.isArray(panelData)
      ? panelData.map(panelData =>
          entries(panelData).filter(
            filterValue.bind(null, omitPairWithFalseyValues)
          )
        )
      : [
          entries(panelData).filter(
            filterValue.bind(null, omitPairWithFalseyValues)
          ),
        ];
  }, [panelData, omitPairWithFalseyValues]);

  return (
    <div className={styles.TableWrap}>
      {tables.map((rows, index) => (
        <div className={styles.InfoPanelTableItem}>
          <table key={index} className={styles.InfoPanelTable}>
            <tbody>
              {rows.map(([title, value], index) => {
                return (
                  <tr
                    key={title + index}
                    className={`InfoPanelTableRow__${slug(title, {
                      lower: true,
                    })}`}
                  >
                    <th>
                      <span>{title}</span>
                    </th>
                    <td>{getValue(value)}</td>
                  </tr>
                );
              })}
            </tbody>
          </table>
        </div>
      ))}
    </div>
  );
}

export default function InfoPanel({
  id,
  title = '',
  actionLinks = [],
  panelData = {},
  className,
  omitPairWithFalseyValues = true,
}: InfoPanelProps) {
  return (
    <div id={id} className={classnames(styles.InfoPanel, className)}>
      {!!title && <Heading className={styles.Title}>{title}</Heading>}
      <div
        className={classnames(
          styles.InfoPanelContent,
          !!title &&
            `${styles.InfoPanelContent}__${slug(title, { lower: true })}`,
          slug(title, { lower: true })
        )}
      >
        <InfoPanelTable
          omitPairWithFalseyValues={omitPairWithFalseyValues}
          panelData={panelData}
        />
        <InfoPanelActionLinks actionLinks={actionLinks || []} />
      </div>
    </div>
  );
}

export function InfoPanelCollapsible({
  id,
  title = '',
  className,
  actionLinks = [],
  panelData = {},
  startCollapsed = true,
}: InfoPanelCollapsibleProps) {
  return (
    <SectionCollapsible
      id={`InfoPanelCollapsible-${id}`}
      className={styles.InfoPanelCollapsible}
      title={title}
      isLoading={false}
      hasItems={true}
    >
      <InfoPanel
        id={id}
        className={className}
        actionLinks={actionLinks}
        panelData={panelData}
      />
    </SectionCollapsible>
  );
}
