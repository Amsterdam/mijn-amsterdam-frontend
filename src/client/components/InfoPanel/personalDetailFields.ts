export const personalDetailFields = [
  'voornamen',
  'achternaam',
  'bsn',
  'geboortedatum',
  'geboorteplaats',
  'geboorteland',
  'nationaliteit',
  'straat',
  'plaats',
];
