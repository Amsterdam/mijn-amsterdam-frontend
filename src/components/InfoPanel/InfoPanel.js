import React from 'react';
import styles from './InfoPanel.module.scss';

function InfoPanelActionLinks({ actionLinks }) {
  return (
    <ul className={styles.InfoPanelActionLinks}>
      {actionLinks.map((actionLink, index) => (
        <li key={`link-${index}`}>
          <a href={actionLink.url}>{actionLink.label}</a>
        </li>
      ))}
    </ul>
  );
}

function InfoPanelTable({ data = {} }) {
  return (
    <table className={styles.InfoPanelTable}>
      <tbody>
        {Object.entries(data).map(([label, value], index) => {
          return (
            <tr key={`row-${index}`}>
              <th>{label}</th>
              <td>{value}</td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
}

export default function InfoPanel({ title = '', actionLinks = [], info = {} }) {
  return (
    <div key={title} className={styles.InfoPanel}>
      {title && <h3 className={styles.InfoPanelTitle}>{title}</h3>}
      <div className={styles.InfoPanelContent}>
        <InfoPanelTable data={info} />
        {!!actionLinks.length && (
          <InfoPanelActionLinks actionLinks={actionLinks} />
        )}
      </div>
    </div>
  );
}
