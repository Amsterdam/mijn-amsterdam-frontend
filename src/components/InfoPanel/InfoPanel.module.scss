@import 'styles/_global.scss';

.InfoPanel .Title {
  margin-left: 16rem;
  margin-bottom: 2rem;

  @include mq-phone() {
    margin-left: 0;
  }
}

.InfoPanelContent {
  display: flex;
  flex-wrap: wrap;
  border-top: 0.1rem solid $color-neutral-grey2;
  padding-top: 2rem;
}

.InfoPanelCollapsible {
  margin-bottom: 2rem;

  @include select(SectionCollapsible_Title_) {
    @include heading-medium;
    margin-left: 16rem;
    margin-bottom: 2rem;
    transform: translateX(-2.1rem);

    @include mq-phone {
      margin-left: 0;
    }
  }

  @include select(SectionCollapsible_Panel) {
    border-bottom: 0;
    border-top: 0.1rem solid $color-neutral-grey2;

    .InfoPanelContent {
      border-top: 0;
    }

    .InfoPanel {
      margin-bottom: 2rem;
    }
  }
}

.InfoPanelTable {
  width: 59%;
  border-collapse: collapse;

  + .InfoPanelTable {
      margin-top: 2rem;
    }

  tbody > tr {
    > td {
      text-align: left;
      white-space: pre-line;
      padding-bottom: 0.5rem;
    }
    > th {
      width: 16rem;
      text-align: right;
      padding-right: 2rem;
      padding-bottom: 0.5rem;
      color: $color-neutral-grey4;
      font-weight: 200;
    }

    th,
    td {
      vertical-align: top;
    }
  }

  @include mq-phone() {
    display: block;

    tbody,
    thead,
    tr,
    td,
    th {
      display: block;
    }
    tbody > tr {
      > th,
      > td {
        width: auto;
        text-align: left;
        width: auto;
        padding-bottom: 0;
      }

      + tr {
        > th {
          padding-top: 2rem;
        }
      }
    }
  }
}

.InfoPanelActionLinks {
  @include reset-list-style;

  > li {
    margin-bottom: 0.6rem;
  }

  @include mq-phone() {
    margin: 2rem 0 0 0;
  }
}
