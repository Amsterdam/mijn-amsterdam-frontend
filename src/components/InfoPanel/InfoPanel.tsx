import React from 'react';
import styles from './InfoPanel.module.scss';
import Linkd from 'components/Button/Button';
import Heading from 'components/Heading/Heading';
import { Unshaped } from 'App.types';
import { entries } from 'helpers/App';
import classnames from 'classnames';
import slug from 'slug';
import SectionCollapsible from 'components/SectionCollapsible/SectionCollapsible';
import { useSessionStorage } from 'hooks/storage.hook';

export interface ActionLink {
  title: string;
  url: string;
  external?: boolean;
}

export type ActionLinksFormatter =
  | ActionLink[]
  | ((sourceData: any) => ActionLink[]);

export interface InfoPanelActionLinksProps {
  actionLinks: ActionLink[];
}

export interface InfoPanelTableProps {
  panelData: Unshaped;
}

export interface InfoPanelProps {
  id?: string;
  title?: string;
  actionLinks?: ActionLink[];
  panelData: Unshaped | Unshaped[];
  className?: string;
}

export interface InfoPanelCollapsibleProps extends InfoPanelProps {
  id: string;
  startCollapsed?: boolean;
}

function InfoPanelActionLinks({ actionLinks }: InfoPanelActionLinksProps) {
  return (
    <ul className={styles.InfoPanelActionLinks}>
      {actionLinks.map((actionLink, index) => (
        <li key={actionLink.title}>
          <Linkd href={actionLink.url} external={actionLink.external}>
            {actionLink.title}
          </Linkd>
        </li>
      ))}
    </ul>
  );
}

function getValue(value: any) {
  if (Array.isArray(value) || typeof value === 'object') {
    return JSON.stringify(value);
  }
  return value;
}

function InfoPanelTable({ panelData = {} }: InfoPanelTableProps) {
  const tables = Array.isArray(panelData)
    ? panelData.map(panelData =>
        entries(panelData).filter(([, value]) => !!value)
      )
    : [entries(panelData).filter(([, value]) => !!value)];
  return (
    <>
      {tables.map((rows, index) => (
        <table key={index} className={styles.InfoPanelTable}>
          <tbody>
            {rows.map(([title, value], index) => {
              return (
                <tr
                  key={title + index}
                  className={`InfoPanelTableRow__${slug(title, {
                    lower: true,
                  })}`}
                >
                  <th>{title}</th>
                  <td>{getValue(value)}</td>
                </tr>
              );
            })}
          </tbody>
        </table>
      ))}
    </>
  );
}

export default function InfoPanel({
  id,
  title = '',
  actionLinks = [],
  panelData = {},
  className,
}: InfoPanelProps) {
  return (
    <div id={id} className={classnames(styles.InfoPanel, className)}>
      {!!title && <Heading className={styles.Title}>{title}</Heading>}
      <div
        className={classnames(
          styles.InfoPanelContent,
          `${styles.InfoPanelContent}__${slug(title, { lower: true })}`,
          slug(title, { lower: true })
        )}
      >
        <InfoPanelTable panelData={panelData} />
        {!!actionLinks.length && (
          <InfoPanelActionLinks actionLinks={actionLinks} />
        )}
      </div>
    </div>
  );
}

export function InfoPanelCollapsible({
  id,
  title = '',
  className,
  actionLinks = [],
  panelData = {},
  startCollapsed = true,
}: InfoPanelCollapsibleProps) {
  const [isCollapsed, setCollapsed] = useSessionStorage(id, startCollapsed);
  return (
    <SectionCollapsible
      className={styles.InfoPanelCollapsible}
      title={title}
      isLoading={false}
      hasItems={true}
      isCollapsed={isCollapsed}
      onToggleCollapsed={() => setCollapsed(!isCollapsed)}
    >
      <InfoPanel
        id={id}
        className={className}
        actionLinks={actionLinks}
        panelData={panelData}
      />
    </SectionCollapsible>
  );
}
