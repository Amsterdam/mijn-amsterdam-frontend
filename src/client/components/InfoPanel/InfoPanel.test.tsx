import React from 'react';
import { shallow } from 'enzyme';
import InfoPanel from './InfoPanel';
import { BrowserRouter } from 'react-router-dom';

const infoData = {
  foo: 'bar',
  bar: '',
};

const title = 'The InfoPanel';

const actionLinks = [
  {
    title: 'An external actionLink!',
    url: 'http://example.org',
    external: true,
  },
  {
    title: 'An internal actionLink!',
    url: 'http://example.org',
  },
];

describe('InfoPanel', () => {
  it('Renders everything correctly', () => {
    expect(
      shallow(
        <BrowserRouter>
          <InfoPanel
            title={title}
            actionLinks={actionLinks}
            panelData={infoData}
          />
        </BrowserRouter>
      ).html()
    ).toMatchSnapshot();
  });

  it('Doesn`t omit falsey values', () => {
    expect(
      shallow(
        <InfoPanel panelData={infoData} omitPairWithFalseyValues={false} />
      ).html()
    ).toMatchSnapshot();
  });
});
