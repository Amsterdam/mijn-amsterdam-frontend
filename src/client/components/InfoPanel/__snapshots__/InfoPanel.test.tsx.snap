// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoPanel Doesn\`t omit falsey values 1`] = `"<div class=\\"InfoPanel\\"><div class=\\"InfoPanelContent\\"><div class=\\"TableWrap\\"><table class=\\"InfoPanelTable\\"><tbody><tr class=\\"InfoPanelTableRow__foo\\"><th>foo</th><td>bar</td></tr><tr class=\\"InfoPanelTableRow__bar\\"><th>bar</th><td></td></tr></tbody></table></div></div></div>"`;

exports[`InfoPanel Renders everything correctly 1`] = `"<div class=\\"InfoPanel\\"><h3 class=\\"Heading Heading__medium Title\\">The InfoPanel</h3><div class=\\"InfoPanelContent InfoPanelContent__the-infopanel the-infopanel\\"><div class=\\"TableWrap\\"><table class=\\"InfoPanelTable\\"><tbody><tr class=\\"InfoPanelTableRow__foo\\"><th>foo</th><td>bar</td></tr></tbody></table></div><ul class=\\"InfoPanelActionLinks\\"><li><a rel=\\"external noopener noreferrer\\" href=\\"http://example.org\\" class=\\"Button Button__plain Button__lean Linkd\\"><svg aria-hidden=\\"true\\" width=\\"14px\\" height=\\"14px\\" fill=\\"#000\\" class=\\"Icon Icon__left\\">Chevron-Right.svg</svg>An external actionLink!</a></li><li><a class=\\"Button Button__plain Button__lean Linkd\\" href=\\"/http://example.org\\"><svg aria-hidden=\\"true\\" width=\\"14px\\" height=\\"14px\\" fill=\\"#000\\" class=\\"Icon Icon__left\\">Chevron-Right.svg</svg>An internal actionLink!</a></li></ul></div></div>"`;
